x/reward/keeper/msg_server_set_reward.go

```diff
@@ -0,0 +1,89 @@
+package keeper
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	spnerrors "github.com/tendermint/spn/pkg/errors"
+	launchtypes "github.com/tendermint/spn/x/launch/types"
+	profiletypes "github.com/tendermint/spn/x/profile/types"
+	"github.com/tendermint/spn/x/reward/types"
+)
+
+func (k msgServer) SetReward(goCtx context.Context, msg *types.MsgSetReward) (*types.MsgSetRewardResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// determine if the chain exists
+	chain, found := k.launchKeeper.GetChain(ctx, msg.LaunchID)
+	if !found {
+		return nil, sdkerrors.Wrapf(launchtypes.ErrChainNotFound, "%d", msg.LaunchID)
+	}
+	// check coordinator
+	coordByAddr, found := k.profileKeeper.GetCoordinatorByAddress(ctx, msg.Provider)
+	if !found {
+		return nil, sdkerrors.Wrap(profiletypes.ErrCoordAddressNotFound, msg.Provider)
+	}
+	if coordByAddr.CoordinatorID != chain.CoordinatorID {
+		return nil, sdkerrors.Wrapf(types.ErrInvalidCoordinatorID, "%d", coordByAddr.CoordinatorID)
+	}
+	// reward can't be changed once launch is triggered
+	if chain.LaunchTriggered {
+		return nil, sdkerrors.Wrapf(launchtypes.ErrTriggeredLaunch, "%d", msg.LaunchID)
+	}
+
+	provider, err := sdk.AccAddressFromBech32(msg.Provider)
+	if err != nil {
+		return nil, spnerrors.Criticalf("can't parse provider address %s", err.Error())
+	}
+
+	rewardPool, found := k.GetRewardPool(ctx, msg.LaunchID)
+	if !found {
+		// create the reward pool and transfer tokens if not created yet
+		if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, provider, types.ModuleName, msg.Coins); err != nil {
+			return nil, sdkerrors.Wrap(types.ErrInsufficientFunds, err.Error())
+		}
+		rewardPool = types.NewRewardPool(msg.LaunchID, ctx.BlockHeight())
+	} else if err := SetBalance(ctx, k.accountKeeper, k.bankKeeper, provider, msg.Coins); err != nil {
+		return nil, spnerrors.Criticalf("can't set balance %s", err.Error())
```

Comment:
We should have a single `spnerrors.Criticalf` at the top of the call stack. Therefore `SetBalance` should not return a critical error. We should parse here the returned error from `SetBalance` and return a critical error here if it is critical